pkgs/development/compilers/factor-lang/mk-factor-application.nix

```diff
@@ -0,0 +1,115 @@
+{ lib
+, writeText
+, makeWrapper
+, factor-lang
+, factor-no-gui
+, librsvg
+, gdk-pixbuf
+} @ initAttrs:
+
+{ name ? "${attrs.pname}-${attrs.version}"
+
+, factor-lang ? if enableUI then initAttrs.factor-lang else initAttrs.factor-no-gui
+
+, stdenv ? factor-lang.stdenv
+
+, nativeBuildInputs ? []
+
+, buildInputs ? []
+
+# Allow overriding the deployed vocabulary name. A $vocabName.factor file must exist!
+, vocabName ? attrs.pname or name
+
+# Allow overriding the binary name
+, binName ? lib.last (lib.splitString "/" vocabName)
+
+# Extra libraries needed, defaults to standard libraries, set [] to disable
+, extraLibs ? with factor-lang; defaultLibs ++ runtimeLibs
+
+# Extra binaries in PATH, defaults to standard paths, set [] to disable
+, extraPaths ? with factor-lang; defaultBins ++ binPackages
+
+, enableUI ? false
+
+, meta ? {}
+
+, ...
+} @ attrs:
+let
+  deployScript = writeText "deploy-me.factor" ''
+    USING: command-line io io.backend io.pathnames kernel namespaces sequences
+    tools.deploy tools.deploy.config tools.deploy.backend vocabs.loader ;
+
+    IN: deploy-me
+
+    : load-and-deploy ( path/vocab -- )
+        normalize-path [
+            parent-directory add-vocab-root
+        ] [
+            file-name dup reload deploy
+        ] bi ;
+
+    : deploy-vocab ( path/vocab path/target -- )
+        normalize-path deploy-directory set
+        f open-directory-after-deploy? set
+        load-and-deploy ;
+
+    : deploy-me ( -- )
+        command-line get dup length 2 = [
+            first2 deploy-vocab
+        ] [
+            drop
+            "usage: deploy-me <PATH-TO-VOCAB> <TARGET-DIR>" print
+            nl
+        ] if ;
+
+    MAIN: deploy-me
+  '';
+in stdenv.mkDerivation ((builtins.removeAttrs attrs [
+  "stdenv" "buildPhase" "installPhase"
+]) // {
```

Comment:
@ShamrockLee suggests not using this patter for new build helpers (builders) [here][1].  I feel the same.

`stdenv` can be renamed.  `buildPhase` and `installPhase` can be just deleted.

[1]: https://github.com/NixOS/nixpkgs/pull/234651#issuecomment-2012342972

Reply:
To allow `mkFactorApplication` to take `finalAttrs: { ... }`, we need to remove

```
(builtins.removeAttrs attrs [ "stdenv" "buildPhase" "installPhase" ]) //
```

And adopt the `overrideAttrs`-based pattern.

```nix
fpArgs:
(stdenv.mkDerivation fpArgs).overrideAttrs (
  finalAttrs:
  {
    ...
  }@attrs:
  {
    ...
  }
)
```

In the current implementation, you already overwrite `buildPhase` and `installPhase`, making removing them from `attrs` unnecessary before passing them down to `stdenv.mkDerivaiton`.

Reply:
Thanks! That is a much nicer pattern. I updated the file.